apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts

```diff
@@ -3,16 +3,60 @@ import asyncHandler from 'express-async-handler';
 import { OpenAI } from 'openai';
 
 import * as error from '@prairielearn/error';
+import { loadSqlEquiv, queryRow } from '@prairielearn/postgres';
 
 import { config } from '../../../lib/config.js';
+import { QuestionSchema } from '../../../lib/db-types.js';
+import { QuestionAddEditor } from '../../../lib/editors.js';
 import { features } from '../../../lib/features/index.js';
-import { generateQuestion } from '../../lib/aiQuestionGeneration.js';
+import { idsEqual } from '../../../lib/id.js';
+import { selectJobsByJobSequenceId } from '../../../lib/server-jobs.js';
+import { generateQuestion, regenerateQuestion } from '../../lib/aiQuestionGeneration.js';
 import { syncContextDocuments } from '../../lib/contextEmbeddings.js';
 
-import { AiGeneratePage } from './instructorAiGenerateQuestion.html.js';
+import { AiGeneratePage, GenerationResults } from './instructorAiGenerateQuestion.html.js';
 
 const router = express.Router();
 
+const sql = loadSqlEquiv(import.meta.url);
+
+export async function saveGeneratedQuestion(
+  res,
+  htmlFileContents: string | undefined,
+  pythonFileContents: string | undefined,
+): Promise<{ qid?: string; redirect?: string }> {
+  const files = {};
+
+  if (htmlFileContents) {
+    files['question.html'] = htmlFileContents;
+  }
+
+  if (pythonFileContents) {
+    files['server.py'] = pythonFileContents;
+  }
+
+  const editor = new QuestionAddEditor({
+    locals: res.locals,
+    files,
+  });
+
+  const serverJob = await editor.prepareServerJob();
+
+  try {
+    await editor.executeWithServerJob(serverJob);
+  } catch (err) {
+    return { redirect: res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId };
```

Comment:
Another option for control flow here: we have support for "throwing" redirects, like you can see here:

https://github.com/PrairieLearn/PrairieLearn/blob/ee7b9af43448e50ffbdd221bcca595c90d80d19a/apps/prairielearn/src/ee/lib/terms.ts#L55

This can keep the caller oblivious of the failure state here, which can be marginally cleaner.

Reply:
Changed in 6919d9f4829f12ab87c64d40f2dccf405a98d563.